miniscope_io/stream_daq.py

```diff
@@ -334,11 +352,24 @@ def _buffer_to_frame(
             Input buffer queue.
         frame_buffer_queue : multiprocessing.Queue[ndarray]
             Output frame queue.
+        metadata_queue : multiprocessing.Queue[ndarray]
+            Queue for storing received metadata.
+        metadata : Optional[Path], optional
+            Path to save metadata, by default None.
+        show_metadata : Optional[bool], optional
+            Whether to display metadata, by default False.
+            Not implemented yet and  itactually might be better to have this in the main process
         """
         locallogs = init_logger("streamDaq.buffer")
 
         cur_fm_num = -1  # Frame number
 
+        if metadata:
+            buffered_writer = BufferedCSVWriter(
```

Comment:
eventually i think we should rewrite some of these perf-sensitive routines as rust or C extension modules, so as much as we can split them up into pure functions the better off we'll be down the line i think

Reply:
we will want the notion of priority among sinks, and that would let us simplify the structure further: what is the next method in the `Queue` pipeline if not a very high priority sink?

Reply:
Seeing this here gave me a better understanding of your earlier Slack message. I can conceptually understand, but I am unsure if I correctly understand your intention at the implementation level.
- How small should we make the phases? I chunked these in a way that initially felt intuitive to me (queue input--queue output), and they still feel kind of okay. So, I'm not exactly sure what function sizes make sense, but I am curious.
- Dispatching tasks to many small functions sounds like having more processes, but do we want to do this? I thought you were initially worried about overheads for handing data over among processes. Maybe the best way is to do profiling, but I wonder if it's likely to improve performance.
- Making a `sink` class including stuff makes sense too.

Reply:
Btw, are there some good references for these styling/structuring topics except for looking into numerous oss? For a non-programmer like me, it's like, "I can follow coding rules and make stuff work, but what's a good size of a code chunk (beyond defining I/Os that make sense)? Putting stuff in a `sink` class makes sense, but doesn't stuff that goes into there makes sense to also be in `io`? What's too many top-level modules? etc." It seems like something to think about because otherwise, a lot of review labor will be focused on you, which could be better.

Reply:
basically a lot of it is vibes, but there are some rules of thumb like "functions/methods should be no longer than 2-dozen lines or so" and "classes, functions, and methods should do one thing" - in this case i think it's mostly a problem when there's ambiguity about where things should go or where to find them. I think `io` is probably too generic of a module for a package named `miniscope-io` since hopefully *all* of it is `io` lol. so in this case refactoring for clarity and DRY